radixdlt/src/main/java/com/radixdlt/consensus/liveness/WeightedRotatingLeaders.java

```diff
@@ -0,0 +1,174 @@
+		private final ValidatorSet validatorSet;
+		private final Comparator<Entry<Validator, UInt384>> weightsComparator;
+		private final Map<Validator, UInt384> weights;
+		private final Validator[] cache;
+		private final Long lcm;
+		private View curView;
+
+		private CachingNextLeaderComputer(ValidatorSet validatorSet, Comparator<Entry<Validator, UInt384>> weightsComparator, int cacheSize) {
+			this.validatorSet = validatorSet;
+			this.weightsComparator = weightsComparator;
+			this.weights = new HashMap<>();
+			this.cache = new Validator[cacheSize];
+
+			UInt256[] powerArray = validatorSet.getValidators().stream().map(Validator::getPower).toArray(UInt256[]::new);
+			// after cappedLCM is executed, the following invariant will be true:
+			// (lcm > 0 && lcm < 2^63 -1 ) || lcm == null
+			// This is due to use of 2^63 - 1 cap and also the invariant from ValidatorSet
+			// that powerArray will always be non-zero
+			UInt256 lcm256 = MathUtils.cappedLCM(UInt256.from(Long.MAX_VALUE), powerArray);
+			this.lcm = lcm256 == null ? null : lcm256.getLow().getLow();
+
+			this.resetToView(View.of(0));
+		}
+
+		private Validator computeHeaviest() {
+			final Entry<Validator, UInt384> max = weights.entrySet().stream()
+				.max(weightsComparator)
+				.orElseThrow(() -> new IllegalStateException("Weights cannot be empty"));
+			return max.getKey();
+		}
+
+		private void computeNext() {
+			// Reset current leader by subtracting total power
+			final int curIndex = (int) (this.curView.number() % cache.length);
+			final Validator curLeader = cache[curIndex];
+			weights.merge(curLeader, UInt384.from(validatorSet.getTotalPower()), UInt384::subtract);
+
+			// Add weights relative to each validator's power
+			for (Validator validator : validatorSet.getValidators()) {
+				weights.merge(validator, UInt384.from(validator.getPower()), UInt384::add);
+			}
+
+			// Compute next leader by getting heaviest validator
+			this.curView = this.curView.next();
+			int index = (int) (this.curView.number() % cache.length);
+			cache[index] = computeHeaviest();
+		}
+
+		private Validator checkCacheForProposer(View view) {
+			if (view.compareTo(curView) <= 0 && view.number() > curView.number() - cache.length) {
+				final int index = (int) (view.number() % cache.length);
+				return cache[index];
+			}
+
+			return null;
+		}
+
+		private void computeToView(View view) {
+			while (view.compareTo(curView) > 0) {
+				computeNext();
+			}
+		}
+
+		private Validator resetToView(View view) {
+			// reset if view isn't in cache
+			if (curView == null || view.number() < curView.number() - cache.length) {
+				if (lcm == null || lcm > view.number()) {
+					curView = View.genesis();
+				} else {
+					long multipleOfLCM = view.number() / lcm;
+					curView = View.of(multipleOfLCM * lcm);
+				}
+
+				for (Validator validator : validatorSet.getValidators()) {
+					weights.put(validator, UInt384.from(UInt128.ONE, UInt256.ZERO).subtract(validator.getPower()));
```

Comment:
Would be better as a constant, rather than creating a new one only to throw away immediately.
```
private static final UInt384 POW_2_256 = UInt384.from(UInt256.MAX_VALUE).increment();
```